File: 127_WordLadder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        wordSet = set(wordList)
        if endWord not in wordSet:
            return 0
        
        q = [endWord]
        dist = 2
        while q:
            newQ = list()
            for candidate in q:
                if candidate == beginWord:
                    return dist
                dist += 1
                wordSet.discard(candidate)     
                for i in xrange(len(candidate)):
                    for j in xrange(26):
                        choice = candidate[:i] + chr(ord('a') + j) + candidate[i + 1:]
                        if choice in wordSet:
                            newQ.append(choice)
            q = newQ
        return 0
		

###A deque solution		
from collections import deque
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        
        def construct_dict(word_list):
            d = {}
            for word in word_list:
                for i in range(len(word)):
                    s = word[:i] + "_" + word[i+1:]
                    d[s] = d.get(s, []) + [word]
            return d
            
        def bfs_words(begin, end, dict_words):
            queue, visited = deque([(begin, 1)]), set()
            while queue:
                word, steps = queue.popleft()
                if word not in visited:
                    visited.add(word)
                    if word == end:
                        return steps
                    for i in range(len(word)):
                        s = word[:i] + "_" + word[i+1:]
                        neigh_words = dict_words.get(s, [])
                        for neigh in neigh_words:
                            if neigh not in visited:
                                queue.append((neigh, steps + 1))
            return 0
        
        d = construct_dict(wordList | set([beginWord, endWord]))
        return bfs_words(beginWord, endWord, d)
                

File: 127_WordLadder.cpp
//62.2%
//In this way, it's too complex
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int step(0);
		unordered_set<string> traced;
		//Check if 'endWord' within wordList
		for(;step < wordList.size(); ++step)
			if(wordList[step] == endWord)
				break;
		if(step == wordList.size())
			return 0;
		
		step = 0;
		vector<string> candidates(1, endWord);
		traced.insert(endWord);
		while(!candidates.empty()){
			++step;
			for(auto& candidate : candidates){
				if(diffWord(candidate, beginWord) == 1)
					return step + 1;
			}
			vector<string> newCandidates;
			for(auto& word : wordList){
				if(traced.find(word) == traced.end()){
					for(auto& candidate : candidates){
						if(diffWord(candidate, word) == 1){
							newCandidates.push_back(word);
							traced.insert(word);
							break;
						}
					}
				}
			}
			candidates = newCandidates;
		}
		return 0;
    }
private:
	//cnt: different letters' number
	int diffWord(const string& firstWord, const string& secondWord){
		int cnt(0);
		for(int i = 0; i < firstWord.size(); ++i){
			if(firstWord[i] != secondWord[i]){
				if(!cnt)
					++cnt;
				else
					return 2;
			}
		}
		return cnt;
	}
};

//It's a typical use case with 'BFS'
//If we're using 'unordered_set' data structure on wordDict
class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordDict) {
        wordDict.insert(endWord);
        queue<string> toVisit;
        addNextWords(beginWord, wordDict, toVisit);
        int dist = 2;
        while (!toVisit.empty()) {
            int num = toVisit.size();
            for (int i = 0; i < num; i++) {
                string word = toVisit.front();
                toVisit.pop();
                if (word == endWord) return dist;
                addNextWords(word, wordDict, toVisit);
            }
            dist++;
        }
		return 0;
    }
private:
    void addNextWords(string word, unordered_set<string>& wordDict, queue<string>& toVisit) {
        wordDict.erase(word);
        for (int p = 0; p < (int)word.length(); p++) {
            char letter = word[p];
            for (int k = 0; k < 26; k++) { 
                word[p] = 'a' + k;
                if (wordDict.find(word) != wordDict.end()) {
                    toVisit.push(word);
                    wordDict.erase(word);
                }
            }
            word[p] = letter;
        } 
    } 
};


//A two ends starting search problem to speed up the whole programme
class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordDict) {
        unordered_set<string> head, tail, *phead, *ptail;
        head.insert(beginWord);
        tail.insert(endWord);
        int dist = 2;
        while (!head.empty() && !tail.empty()) {
            if (head.size() < tail.size()) {
                phead = &head;
                ptail = &tail;
            }
            else {
                phead = &tail; 
                ptail = &head;
            }
            unordered_set<string> temp; 
            for (auto itr = phead -> begin(); itr != phead -> end(); itr++) {
                string word = *itr;
                wordDict.erase(word);
                for (int p = 0; p < (int)word.length(); p++) {
                    char letter = word[p];
                    for (int k = 0; k < 26; k++) {
                        word[p] = 'a' + k;
                        if (ptail -> find(word) != ptail -> end())
                            return dist;
                        if (wordDict.find(word) != wordDict.end()) {
                            temp.insert(word);
                            wordDict.erase(word);
                        }
                    }
                    word[p] = letter;
                }
            }
            dist++;
            swap(*phead, temp);
        }
        return 0; 
    }
};